diag_manager/fms_diag_buffer.F90

```diff
@@ -0,0 +1,1561 @@
+  class(*),intent(in) :: buff_type !< allocates to the type of buff_type
+  integer, intent(in) :: buff_size !< dimension bounds
+  integer, optional :: diurnal_samples !< number of diurnal samples, passed in from diag_yaml
+  integer :: n_samples = 1
+
+  if(present(diurnal_samples)) then
+    n_samples = diurnal_samples
+  endif
+
+  if(allocated(this%buffer)) call mpp_error(FATAL, "allocate_buffer_1d: buffer already allocated")
+  select type (buff_type)
+    type is (integer(kind=i4_kind))
+      allocate(integer(kind=i4_kind) :: this%buffer(buff_size))
+      allocate(integer(kind=i4_kind) :: this%counter(buff_size))
+      allocate(integer(kind=i4_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%typestr = 'i4'
+    type is (integer(kind=i8_kind))
+      allocate(integer(kind=i8_kind) :: this%buffer(buff_size))
+      allocate(integer(kind=i8_kind) :: this%counter(buff_size))
+      allocate(integer(kind=i8_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%typestr = 'i8'
+    type is (real(kind=r4_kind))
+      allocate(real(kind=r4_kind) :: this%buffer(buff_size))
+      allocate(real(kind=r4_kind) :: this%count_0d(buff_size))
+      allocate(real(kind=r4_kind) :: this%counter(n_samples))
+      this%counter = 0
+      this%typestr = 'r4'
+    type is (real(kind=r8_kind))
+      allocate(real(kind=r8_kind) :: this%buffer(buff_size))
+      allocate(real(kind=r8_kind) :: this%count_0d(buff_size))
+      allocate(real(kind=r8_kind) :: this%counter(n_samples))
+      this%counter = 0
+      this%typestr = 'r8'
+    class default
+       call mpp_error("allocate_buffer_1d", &
+           "The buff_type value passed to allocate a buffer is not a r8, r4, i8, or i4",&
+           FATAL)
+  end select
+
+  allocate(this%num_elements(n_samples))
+  allocate(this%buffer_dims(1))
+  this%num_elements = 0
+  this%count_0d   = 0
+  this%buffer_dims(1) = buff_size
+
+end subroutine allocate_buffer_1d
+!> allocates a 2D buffer to given buff_type type
+!! TODO fails with gnu
+subroutine allocate_buffer_2d(this, buff_type, buff_sizes, diurnal_samples)
+  class(buffer2d), intent(inout), target :: this !< 2D buffer object
+  class(*),intent(in) :: buff_type !< allocates to the type of buff_type
+  integer, intent(in) :: buff_sizes(2) !< dimension sizes
+  integer, optional :: diurnal_samples !< number of diurnal samples, passed in from diag_yaml
+  integer :: n_samples = 1
+
+  if(present(diurnal_samples)) then
+    n_samples = diurnal_samples
+  endif
+
+  if(allocated(this%buffer)) call mpp_error(FATAL, "allocate_buffer_2d: buffer already allocated")
+  select type (buff_type)
+    type is (integer(kind=i4_kind))
+      allocate(integer(kind=i4_kind) :: this%buffer(buff_sizes(1), buff_sizes(2)))
+      allocate(integer(kind=i4_kind) :: this%counter(buff_sizes(1), buff_sizes(2)))
+      allocate(integer(kind=i4_kind) :: this%count_0d(n_samples))
+      this%counter = 0
+      this%typestr = 'i4'
```

Comment:
Depending on how this will be used, maybe we should store it as a parameter integer:
https://github.com/NOAA-GFDL/FMS/blob/e8870d4fdcb97388119fa97a2c93a9cdfcc86fc3/diag_manager/diag_data.F90#L84-L87

Reply:
Yeah that would be good but it might end up being removed anyway. I noticed there's another `vartype`  in the field object so we probably don't need both.